packages/settingregistry/src/settingregistry.ts

```diff
@@ -1478,21 +1498,81 @@ namespace Private {
-        const reified =
-          (reifyDefault(props, undefined, definitions) as PartialJSONObject) ??
-          {};
-        for (const prop in reified) {
-          if ((result[item] as PartialJSONObject)?.[prop]) {
-            reified[prop] = (result[item] as PartialJSONObject)[prop];
+        if (props.type === 'object') {
+          // Use the values that are hard-coded in the default array over the defaults for each field.
+          const reified =
+            (reifyDefault(
+              props,
+              undefined,
+              definitions
+            ) as PartialJSONObject) ??
+            result[item] ??
+            {};
+          for (const prop in reified) {
+            if ((result[item] as PartialJSONObject)?.[prop]) {
+              reified[prop] = (result[item] as PartialJSONObject)[prop];
+            }
           }
+          result[item] = reified;
         }
-        result[item] = reified;
       }
 
       return result;
     } else {
       return schema.default;
     }
   }
+
+  /**
+   * Upgrade shortcuts to ensure no breaking changes between minor versions.
+   */
+  export function upgradeShortcuts(
+    shortcuts: ISettingRegistry.IShortcut[]
+  ): ISettingRegistry.IShortcut[] {
+    const selectorDeprecationWarnings = new Set();
+    const changes = [
+      {
+        old: '.jp-Notebook:focus.jp-mod-commandMode',
+        new: '.jp-Notebook.jp-mod-commandMode :focus:not(:read-write)',
+        versionDeprecated: 'JupyterLab 4.1'
+      },
+      {
+        old: '.jp-Notebook:focus',
+        new: '.jp-Notebook.jp-mod-commandMode :focus:not(:read-write)',
+        versionDeprecated: 'JupyterLab 4.1'
+      },
+      {
+        old: '[data-jp-traversable]:focus',
+        new: '.jp-Notebook.jp-mod-commandMode :focus:not(:read-write)',
+        versionDeprecated: 'JupyterLab 4.1'
+      },
+      {
+        old: '[data-jp-kernel-user]:focus',
+        new: '[data-jp-kernel-user] :focus:not(:read-write)',
+        versionDeprecated: 'JupyterLab 4.1'
+      }
+    ];
+    const upgraded = shortcuts.map(shortcut => {
+      const oldSelector = shortcut.selector;
+      let newSelector = oldSelector;
+      for (const change of changes) {
+        if (oldSelector.includes(change.old)) {
+          newSelector = oldSelector.replace(change.old, change.new);
+          selectorDeprecationWarnings.add(
+            `"${change.old}" was replaced with "${change.new}" in ${change.versionDeprecated}`
+          );
+        }
+      }
+      shortcut.selector = newSelector;
+      return shortcut;
+    });
+    if (selectorDeprecationWarnings.size > 0) {
+      console.warn(
+        'Deprecated shortcut selectors: ' +
+          [...selectorDeprecationWarnings].concat('\n') +
+          '\n\nThe selectors will be substituted transparently this time, but need to be updated at source before next major release.'
+      );
```

Comment:
It should be nice to have a way to warn only once, when all the plugins are loaded.
This is not for this PR, but having a way to concatenate all messages from plugin loading should be nice, to avoid duplication.
For example, if we restore only one selector from `plugin.json`, we receive 16 warnings in the console.

Reply:
Indeed. We could use some global variable in `Private` to de-duplicate; let me take a look.